redis/cluster.py

```diff
@@ -2093,21 +2145,107 @@ def __bool__(self):
         "Pipeline instances should  always evaluate to True on Python 3+"
         return True
 
+    def _get_client_and_connection_for_transaction(self) -> Tuple[Redis, Connection]:
+        """
+        Find a connection for a pipeline transaction.
+
+        For running an atomic transaction, watch keys ensure that contents have not been
+        altered as long as the watch commands for those keys were sent over the same
+        connection. So once we start watching a key, we fetch a connection to the
+        node that owns that slot and reuse it.
+        """
+        if not self.pipeline_slots:
+            raise RedisClusterException(
+                "At least a command with a key is needed to identify a node"
+            )
+
+        node: ClusterNode = self.nodes_manager.get_node_from_slot(
+            list(self.pipeline_slots)[0], False
+        )
+        redis_node: Redis = self.get_redis_connection(node)
+        if self.transaction_connection:
+            if not redis_node.connection_pool.owns_connection(
+                self.transaction_connection
+            ):
+                previous_node = self.nodes_manager.find_connection_owner(
+                    self.transaction_connection
+                )
+                previous_node.connection_pool.release(self.transaction_connection)
+                self.transaction_connection = None
+
+        if not self.transaction_connection:
+            self.transaction_connection = get_connection(redis_node, ("INFO",))
+
+        return redis_node, self.transaction_connection
+
     def execute_command(self, *args, **kwargs):
         """
         Wrapper function for pipeline_execute_command
         """
-        return self.pipeline_execute_command(*args, **kwargs)
+        slot_number: Optional[int] = None
+        if args[0] not in self.NO_SLOTS_COMMANDS:
+            slot_number = self.determine_slot(*args)
+
+        if (
+            self.watching or args[0] in self.IMMEDIATE_EXECUTE_COMMANDS
+        ) and not self.explicit_transaction:
+            if args[0] == "WATCH":
+                self._validate_watch()
+
+            if slot_number is not None:
+                if self.pipeline_slots and slot_number not in self.pipeline_slots:
```

Comment:
Looks a bit confusing to me, transaction should be exactly executed against single hash slot, why do we check for a set of slots?